pkg/cli/codearea.go

```diff
@@ -196,6 +200,56 @@ func (w *codeArea) handlePasteSetting(start bool) bool {
 	return true
 }
 
+// Try to expand a non-word abbreviation.
+func (w *codeArea) abbrExpansion() {
+	var abbr, full string
+	// Try to expand an abbreviation, preferring the longest one
+	w.Abbreviations(func(a, f string) bool {
+		if strings.HasSuffix(w.inserts, a) && len(a) > len(abbr) {
+			abbr, full = a, f
+		}
+		return false // check the next abbr
+	})
+	if len(abbr) > 0 {
+		c := &w.State.Buffer
+		*c = CodeBuffer{
+			Content: c.Content[:c.Dot-len(abbr)] + full + c.Content[c.Dot:],
+			Dot:     c.Dot - len(abbr) + len(full),
+		}
+		w.resetInserts()
+	}
+}
+
+// Try to expand a word abbreviation.
+func (w *codeArea) wordAbbrExpansion() {
+	w.StateMutex.Lock()
+	defer w.StateMutex.Unlock()
+
+	ilen := len(w.inserts)
+	w.WordAbbreviations(func(a, f string) bool {
+		alen := len(a)
+		if alen > ilen { // skip abbr since it can't match the insert buffer
+			return false // check the next abbr
+		}
+		if ilen > alen {
+			rune, _ := utf8.DecodeLastRuneInString(w.inserts[:ilen-alen])
+			if unicode.IsLetter(rune) || unicode.IsDigit(rune) {
```

Comment:
Elvish's editor has a concept of *word* that is a sequence of non-space characters (space is defined by [unicode.IsSpace](https://godoc.org/unicode#IsSpace)).

It also has a concept of *small word* that is a sequence of letters or numbers (as defined by unicode.IsLetter and unicode.IsNumber).

It would be useful to align the concept of "word abbreviations" to one of those.

Reply:
Small word is what we want here so I'll replace IsDigit() with IsNumber() and add an inline comment.

P.S., Should this use of `IsDigit()` be replaced with `IsNumber()`?

https://github.com/elves/elvish/blob/05156ea239d6c902f7ec10fe6ab74fd4a26c734d/cmd/examples/cli/main.go#L20